### PR TITLE
[system test] [doc] bridge package 

### PR DESCRIPTION
**Pull request by see-quick**
### Type of change

- Enhancement / new feature
- Refactoring
- Documentation

### Description

This PR adds documentation for the Bridge package in our tests. I would like to split this into more minor PRs for easier reviews, rather than opening one big PR for all Suites within STs.  

### Checklist

- [x] Make sure all tests pass
- [x] Update documentation

### CODE REVIEW

File: systemtest/pom.xml [hunk 1]
```diff
@@ -233,6 +233,12 @@
             <groupId>io.skodjob</groupId>
             <artifactId>test-frame-metrics-collector</artifactId>
         </dependency>
+        <dependency>
+            <groupId>io.skodjob</groupId>
+            <artifactId>test-docs-generator-maven-plugin</artifactId>
+            <version>${skodjob-doc.version}</version>
+            <scope>test</scope>
+        </dependency>
```
Comment (scholzj) [inline systemtest/pom.xml:241]: This should be defined in the main pom for dependency management

File: systemtest/pom.xml [hunk 2]
```diff
@@ -291,8 +297,35 @@
                             </ignoredUnusedDeclaredDependencies>
                         </configuration>
                     </execution>
+                    <execution>
+                        <phase>generate-sources</phase>
+                        <goals>
+                            <goal>copy-dependencies</goal>
+                        </goals>
+                        <configuration>
+                            <outputDirectory>${project.build.directory}/lib</outputDirectory>
```
Comment (scholzj) [inline systemtest/pom.xml:306]: Why do you need this?

File: systemtest/pom.xml [hunk 3]
```diff
@@ -291,8 +297,35 @@
...
+                    <execution>
+                        <phase>post-integration-test</phase>
+                        <goals>
+                            <goal>test-docs-generator</goal>
+                        </goals>
+                    </execution>
+                </executions>
+                <configuration>
+                    <testsPath>${project.basedir}/src/test/java/io/strimzi/systemtest</testsPath>
+                    <docsPath>${project.basedir}/docs/</docsPath>
```
Comment (scholzj) [inline systemtest/pom.xml:325]: Why `/docs/`? Yuo should probably explain a bit more how this will work.

File: systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeST.java [hunk 4]
```diff
@@ -63,12 +70,63 @@
 @Tag(REGRESSION)
 @Tag(BRIDGE)
 @Tag(INTERNAL_CLIENTS_USED)
+@SuiteDoc(
+    description = @Desc("Test suite for various Kafka Bridge operations."),
+    contact = @Contact(name = "Lukas Kral", email = "lukywill16@gmail.com"),
```
Comment (scholzj) [inline systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeST.java:75]: I don't think this makes sense. You do not want to have tests tied to a particular person.

File: systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeCorsST.java [hunk 5]
```diff
@@ -53,6 +76,25 @@ public class HttpBridgeCorsST extends AbstractST {
...
+        steps = {
+            @Step(value = "Set up the Kafka Bridge user and configuration", expected = "Kafka Bridge user and configuration are set up"),
+            @Step(value = "Construct the request URL and headers", expected = "URL and headers are constructed properly"),
+            @Step(value = "Send OPTIONS request to Kafka Bridge and capture the response", expected = "Response is captured from Bridge"),
+            @Step(value = "Validate the response contains expected status codes and headers", expected = "Response has correct status codes and headers for allowed origin"),
+            @Step(value = "Send GET request to Kafka Bridge and capture the response", expected = "Response is captured from Bridge for GET request"),
+            @Step(value = "Check if the GET request response is '404 Not Found'", expected = "Response for GET request is 404 Not Found")
+        },
+        useCases = {
+            @UseCase(id = "validate_cors_handling_allowed_origin")
```
Comment (scholzj) [inline systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeCorsST.java:91]: Where are these defined?

File: systemtest/pom.xml (hunk 2 above)
Comment (see-quick) [reply systemtest/pom.xml:306]: Well I did not write the plugin but here is the answer from the `README.md` [1] from repo
```
maven-dependency-plugin is needed for proper loading the classes. Without it the plugin will return NoClassDefFound exception and will fail.
```
Maybe @im-konge  or @Frawless could answer better? 

[1] - https://github.com/skodjob/test-metadata-generator

Comment (scholzj) [reply systemtest/pom.xml:306]: Seems like some bug or something? We do not seem to need it for generating the CRDs or docs.

File: systemtest/pom.xml (hunk 3 above)
Comment (see-quick) [reply systemtest/pom.xml:325]: This is basically the main consideration I have: whether we should go for:

1. Generating documentation in `/systemtest/docs`, where one could then review how our system tests function from the documentation. This documentation would always be synced with the code docs, i.e., @SuiteDoc and @TestDoc. The primary purpose is to provide a source of truth between developers and testers during test triage.
2. The second approach would be to consider other options for where we could generate this documentation.
3. The third option is to let end-users generate such documentation.

and for me it was the first approach, which gave me the most sense...but that's not only up to me so maybe:
@strimzi/system-test-contributors have better idea ;)

File: systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeST.java (hunk 4 above)
Comment (see-quick) [reply systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeST.java:75]: So having a `@Contact` does not make sense for you right? Do you mean remove it completely or maybe re-place with multiple responsible person?

File: systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeCorsST.java (hunk 5 above)
Comment (see-quick) [reply systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeCorsST.java:91]: Should we have them defined somewhere? I tried to use the use case like having the closer description of the test or suite. I assume it could help if we have a detailed description of such use cases but not sure how that helps ;) 

 ^^ @strimzi/system-test-contributors

File: systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeST.java (hunk 4 above)
Comment (scholzj) [reply systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeST.java:75]: It is a community, people come and go, and so on. So if someone has any questions, they should come to GitHub Discussions, Slack, or Mailing list. Not to the contact person. If you need some tracking, I would maybe consider having something like `@Authors`. But I would try it as unrelated to any particular person(s) as possible. I would probably also use only GitHub handles or something rather than email.

But as it is right now - with `@Contact` and `email`, it is just as if you are telling someone who tries to run it to send Lukas an email with some questions.

File: systemtest/pom.xml (hunk 3 above)
Comment (scholzj) [reply systemtest/pom.xml:323]: Would `developer-docs/...` be a better place to have it? There are already docs for how to test etc. In the system tests it would be completely lost. If you plan to have it committed (which for me is fine), you should have an azure pipelines check for the files to be up to date similarly to what we have for `packaging/`.

File: systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeCorsST.java (hunk 5 above)
Comment (scholzj) [reply systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeCorsST.java:91]: It seems weird to have it as `id` with snake case if it doesn't link to anything. Would just some freetext `description` be better?

File: systemtest/pom.xml (hunk 3 above)
Comment (im-konge) [reply systemtest/pom.xml:323]: As I mentioned offline, I'm not sure if it should be part of the operators repository, as it can be a lot of lines and files that you would pull -> and I mentioned that maybe it can be stored in different repository and we could point to that repository in our README. But if everyone will be fine with having it inside the operators repository, then fine -> it should be in the `developer-docs/...` as Jakub mentioned

File: systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeST.java (hunk 4 above)
Comment (im-konge) [reply systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeST.java:75]: I can imagine having there handle for the "strimzi-systemtest-contributors" group, but that doesn't need to be in every test.
Having something like `@Authors` will be fine, but IMO it will anyway be taken as list of people which should be pinged because the test is failing (so similarly to the `@Contact`)

File: systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeCorsST.java (hunk 5 above)
Comment (im-konge) [reply systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeCorsST.java:91]: The main idea was to create MD files (as it is mentioned in the generator's repo and its README), where you will have some description of the usecase and then list of testcases (with links to particular MD files) underneath.

File: systemtest/pom.xml (hunk 2 above)
Comment (im-konge) [reply systemtest/pom.xml:304]: IIRC it's because the way how the test-classes libs are stored during the build etc. I think that we looked on removing the dependency, but it wasn't possible due to something. I can take a look at it, but it was something around the dependencies availability across build phases.

File: systemtest/pom.xml (hunk 3 above)
Comment (scholzj) [reply systemtest/pom.xml:323]: Separate repo will bring it out of sync, would not allow to track releases, etc. I think the option is to store it in this repo, or not store it and have people generate it when needed from the code.

Comment (im-konge) [reply systemtest/pom.xml:323]: @Frawless @henryZrncik @jankalinic any opinion what could be better (having it in the repo or leaving it for users to generate it)?

Comment (henryZrncik) [reply systemtest/pom.xml:323]: I guess that once we will have all docs merged new changes will not requiere that much changes; but if it seems like a lot letting user generate it seems to me as perfectly good option.

File: systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeST.java (hunk 4 above)
Comment (see-quick) [reply systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeST.java:75]: > Having something like @Authors will be fine, but IMO it will anyway be taken as list of people which should be pinged because the test is failing (so similarly to the @Contact)

So should eliminate the `@Contact` and not use it at all? It seems to be the best option I assume...Option with having an `@Authors` annotation will be very redundant...and I am not sure we want to have all names in all tests but maybe I don't know.... What is your thoughts about this @jankalinic @jankalinic @Frawless ? If there are no any....I would proceed with not using `@Contact`

Comment (Frawless) [reply systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeST.java:75]: If current implementation allow you to remove, feel free to remove it, I am fine with it. The field is used for reporting purposes and atm we don't need it here. In case `Contact` is required, you will have to update it in the generator itself.

Comment (jankalinic) [reply systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeST.java:75]: To me it seems authors or contact is sufficient on it's own. Having strimzi st contributors there instead would be great though.

File: systemtest/pom.xml (hunk 3 above)
Comment (see-quick) [reply systemtest/pom.xml:323]: Okay, so I guess 1. approach (i.e., generating such doc here) and specifically in `developers-docs/` would be the best fit. I will also implement that Azure check which you mentioned Jakub ;)

Comment (Frawless) [reply systemtest/pom.xml:323]: +1 for store it in repo and +1 for checking if the files are in sync. It would be great to have check that the docs are generated for new tests, but that's something we cannot do until all tests will be documented and even after I am not sure how it will work. Guess we can revisit it later.

File: systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeCorsST.java (hunk 5 above)
Comment (see-quick) [reply systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeCorsST.java:88]: Looking back to this...Would it add so much value to define all the use cases? I mean it's a lot of work where you imagine that we would end up documenting whole STs. So now I am thinking about if descriptive use cases like this are not enough for now. And if that would be a problem in future could we add those?

File: systemtest/pom.xml (hunk 3 above)
Comment (see-quick) [reply systemtest/pom.xml]: I have updated this, please check it :)

File: systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeCorsST.java (hunk 5 above)
Comment (scholzj) [reply systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeCorsST.java:88]: Not sure if this was at me, at Lukas, or someone else. My main issue is that you introduce some use-case IDs that do not lead anywhere. So as I said, I can imagine for example:
* Removing it completely
* Changing the `id` to some `description` and make it _one sentence_ description of the use-case

It just doesn't seem make sense to have some use case ID that does not lead anywhere.

File: .azure/scripts/uncommitted-changes.sh
Comment (scholzj) [inline .azure/scripts/uncommitted-changes.sh:1]: You should change the Azure pipeline comment as well.

File: development-docs/systemtest-docs/md/io/strimzi/systemtest/bridge/HttpBridgeCorsST.md
Comment (scholzj) [inline development-docs/systemtest-docs/md/io/strimzi/systemtest/bridge/HttpBridgeCorsST.md:1]: Do we need to have such a deep nesting? Will it be really helpful? I assume that at this point:
* You do not need the `md` part as you have only markdown.
* You do not need to have `docs` twice in the path => so maybe `developer-docs/systemtests/...` would be suffient?
* Would it make sense to name the files `io.strimzi.systemtest.bridge.HttpBridgeCorsST.md`? You would have a shallower directory structure easier to browse and yet keep the information about the package. I guess other option might be to truncate the `io/strimzi/systemtest` from the path?

File: systemtest/pom.xml
Comment (scholzj) [inline systemtest/pom.xml:1]: Is it expected to have all these messages and seemingly errors int he log?

```
[INFO] --- test-docs-generator-maven-plugin:0.1.0:test-docs-generator (default) @ systemtest ---
[INFO] Starting generator
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
Could not create parent directories
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
Could not create parent directories
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
Could not create parent directories
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
Could not create parent directories
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
[INFO] Skipping fmf generation
Could not create parent directories
[INFO] Skipping fmf generation
[INFO] Done
```

File: systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeCorsST.java (hunk 5 above)
Comment (see-quick) [reply systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeCorsST.java:88]: ^^ @strimzi/system-test-contributors any opinion on this?

Comment (im-konge) [reply systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeCorsST.java:88]: I guess it's useless to have use-cases there without any description what it actually covers. Because yeah, you can have one or two which are descriptive enough, but rest which will be a bit confusing when you use just the ID.
The idea of use-cases and linking between is that you will have one file with the description and the test-cases, which cover this use-case.

In case that you think it's a lot of work documenting the STs (which is the main goal of this), you should either remove it or leave it for someone else.

Also I don't think that it's good to have ID as a whole sentence. It's nice that you generate it, but you will end up creating multiple use-cases which are actually just one (and you will have to add description for all of them). So in this case, it makes more sense to go through the tests and add use-cases that are really relevant manually.

Comment (Frawless) [reply systemtest/src/test/java/io/strimzi/systemtest/bridge/HttpBridgeCorsST.java:88]: IIRC the original idea was to have some high-level use-case with some reasonable description - I would say maybe detailed description of interaction between user and Strimzi. For example `User can migrade from Zookeeper-based Kafka cluster to KRaft based cluster` + some details, description of corner cases etc. This description would be something that will be provided as part of planning/propsal? Then the each test cases will be linked to use-cases so each use-case will be covered by N test-cases. 

The idea was also to easily find which tests covers what use-cases/features, etc.

File: development-docs/systemtest-docs/md/io/strimzi/systemtest/bridge/HttpBridgeCorsST.md
Comment (see-quick) [reply development-docs/systemtest-docs/md/io/strimzi/systemtest/bridge/HttpBridgeCorsST.md:1]: >  You do not need the md part as you have only markdown.

I am not sure I can change it. I mean we could implement it inside generator I think that if only one format is provided we would not generate it. 

> You do not need to have docs twice in the path => so maybe developer-docs/systemtests/... would be sufficient?

Sure I would update it.

> Would it make sense to name the files io.strimzi.systemtest.bridge.HttpBridgeCorsST.md? You would have a shallower directory structure easier to browse and yet keep the information about the package. I guess other option might be to truncate the io/strimzi/systemtest from the path?

Well, I am not sure about this. I see your point and think it has some advantages but the current approach follows the structure of the tests, which for me is more natural... WDYT? @im-konge @Frawless

File: systemtest/pom.xml
Comment (see-quick) [reply systemtest/pom.xml:1]: I guess so. I mean having this `Skipping fmf generation` as `[DEBUG]` would be ideal but 🤷 .

Comment (scholzj) [reply systemtest/pom.xml:1]: That is one thing. The other is that it seems to report some errors about directory creation?

Comment (Frawless) [reply systemtest/pom.xml:1]: I will take a look on it in the generator

Comment (see-quick) [reply systemtest/pom.xml:1]: Resolved in new version

File: development-docs/systemtest-docs/md/io/strimzi/systemtest/bridge/HttpBridgeCorsST.md
Comment (see-quick) [reply development-docs/systemtest-docs/md/io/strimzi/systemtest/bridge/HttpBridgeCorsST.md:1]: changed to `io.strimzi.systemtest.bridge.HttpBridgeCorsST.md`

Comment (see-quick) [reply development-docs/systemtest-docs/md/io/strimzi/systemtest/bridge/HttpBridgeCorsST.md:1]: and also now there is no `md` directory

File: development-docs/systemtests/labels/annotation.md [hunk 6]
```diff
@@ -0,0 +1,24 @@
+### Annotation
+
+#### Label: `annotation`
+The `annotation` label is used for tests that verify the correct application and validation of custom annotations in Kubernetes resources. This includes ensuring that custom annotations are properly set during resource creation and correctly retrieved and validated against expected values.
```
Comment (Frawless) [inline development-docs/systemtests/labels/annotation.md:4]: isn't it too low level for label?

File: development-docs/systemtests/labels/annotation.md [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    - Retrieving Kubernetes resources with custom annotations.
+    - Validating that the retrieved annotations match the expected values.
+
+3. **Annotation Validation**:
+    - Filtering specific annotations from resources.
+    - Verifying that filtered annotations align with predefined criteria and expectations.
+
+
+
+
```
Comment (Frawless) [inline development-docs/systemtests/labels/annotation.md:21]: it seems there is a different number of empty lines for each file, maybe it should be united somehow

File: development-docs/systemtests/labels/configuration.md [hunk 8]
```diff
@@ -0,0 +1,30 @@
+# Configuration
+
+### Label Description
+The `configuration` label encompasses all tests related to configuration management. This includes creating, updating, replacing, and dynamically updating configurations across components.
```
Comment (Frawless) [inline development-docs/systemtests/labels/configuration.md:4]:
```suggestion
The `configuration` label encompasses all tests related to configuration management. This includes creating, updating, replacing, and dynamically updating configurations across all operands managed by Strimzi.
```

File: development-docs/systemtests/labels/label.md [hunk 9]
```diff
@@ -0,0 +1,23 @@
+# Label
+
+#### Label: `label`
```
Comment (Frawless) [inline development-docs/systemtests/labels/label.md:3]: I think this might be grouped with `annotation` label?

File: development-docs/systemtests/labels/plaintext.md [hunk 10]
```diff
@@ -0,0 +1,24 @@
+# Plaintext
+
+The `plaintext` label focuses on testing communication mechanisms over unencrypted PLAINTEXT channels.
```
[the remark this answers is not in the document]
Comment (see-quick) [reply development-docs/systemtests/labels/plaintext.md:3]: > Guess it will cover in the end just a several tests as well or?

I mean plaintext will have a lot of tests I think. In other tests, where we use unencrypted messages :)

File: development-docs/systemtests/labels/label.md (hunk 9 above)
Comment (see-quick) [reply development-docs/systemtests/labels/label.md:3]: >  think this might be grouped with annotation label?

Not sure, I think sometimes in tests you use only Labels (f.e., in this test `testLabelsExistenceAndManipulation`) we verify only labels. On the other hand, there are a few tests, where only annotation is used... :)

File: development-docs/systemtests/labels/plaintext.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (see-quick) [reply development-docs/systemtests/labels/plaintext.md]: For now updated to have just one in this package. In other packages we will have more (e.g., upgrade, downgrade, auth, authz, certificate exchange)

**Review by Frawless — APPROVED** [review]
That's exactly how I thought it will looks like. LGTM.

**Review by henryZrncik — APPROVED** [review]

**Review by jankalinic — APPROVED** [review]

**Review by im-konge — APPROVED** [review]

**Review by scholzj — APPROVED** [review]
Nice work. Thanks.
